File: LeetCode/3427-special-array-ii/special-array-ii.java
class Solution {
    public boolean[] isArraySpecial(int[] nums, int[][] queries) {
        int n = nums.length;
        int qn = queries.length;
        boolean[] answer = new boolean[qn];
        List<int[]> falseList = new ArrayList<>();


        for(int i = 0; i < n-1; i++){
            if ((nums[i] % 2 == 0 && nums[i+1] % 2 == 0) || (nums[i] % 2 != 0 && nums[i+1] % 2 != 0)){
                falseList.add(new int[]{i, i+1});
            }
        }
        
        for(int i = 0; i < qn; i++){
            int left = 0;
            int right = falseList.size()-1;
            boolean check = true;

            while (left <= right){
                int mid = (left+right)/2;
                if (queries[i][0] <= falseList.get(mid)[0] && falseList.get(mid)[0] <= queries[i][1] && 
                    queries[i][0] <= falseList.get(mid)[1] && falseList.get(mid)[1] <= queries[i][1]) {
                        check = false;
                        break;
                }
                else if (falseList.get(mid)[1] <= queries[i][0]){
                    left = mid+1;
                }
                else{
                    right = mid-1;
                }
            }
            answer[i] = check;

        }

        return answer;
    }
}

File: 3427-special-array-ii/special-array-ii.py
class Solution:
    def isArraySpecial(self, nums: List[int], queries: List[List[int]]) -> List[bool]:
        n = len(nums)
        false_list = []
        answer = []

        for i in range(n-1):
            if (nums[i] % 2 == 0 and nums[i+1] %2 == 0) or (nums[i] % 2 != 0 and nums[i+1] %2 != 0):
                false_list.append((i, i+1))
        print(false_list)
        false_cnt = len(false_list)

        for query in queries:
            left, right = 0, false_cnt-1
            false_check = True
            while left <= right:
                mid = (left+right)//2
                if query[0] <= false_list[mid][0] <= query[1] and query[0] <= false_list[mid][1] <= query[1]:
                    false_check = False
                    break
                elif false_list[mid][1] <= query[0]:
                    left = mid+1
                else:
                    right = mid-1
            
            answer.append(false_check)
        
        return answer